design/Events_design.md

```diff
@@ -0,0 +1,113 @@
+
+
+#### Problem 3: How to encrypt events before they leave the enclave
+
+#### Problem 4: The mechanism should not leak general information on what activity is happening.
+
+
+## Proposed solution
+
+### Prerequisites
+An event `E1` is emitted after executing transaction `TX1` with a from address `A1`.
+`E1` is formed of multiple topics, 2 of which are addresses (`A2` and `A3`), and there are some other random values.
+
+An event `E2` is emitted by a transaction `TX2` from an address `A4`.
+`E2` is formed of multiple topics, none of which are addresses.
+
+
+### Solution
+
+1. We modify the subscription call to accept a list of owning accounts. 
+Each account must be signed with the latest viewing key, to prevent someone from asking random events, just to leak info. 
+The call will fail if there are no viewing keys for all those accounts.
+This is possible because the subscription call is implemented on Obscuro, and made by the wallet_extension, so it doesn't have to be compatible with Ethereum.
+For our RPC client it would an authenticated subscription.
+
+2. We need to pass into the enclave the "Subscription" objects, and create an rpc streaming for each. 
+
+3. The enclave runs all the usual filtering logic on this event, and determines if there is any subscription that requested it.
+
+4. Then, there is an extra step (inside the enclave as well) to determine whether that party is allowed to view that event. 
+
+This can be done by running an intersection between:
+  - all accounts found in the event topics plus the `from` of the transaction: `E1->(A1,A2,A3)` 
+  - and the accounts from the subscription. 
+
+If there is at least one element, the event is encrypted with the first of the result.
+
+6. For event `E2`, that has no addresses in the topics, there are a couple of options:
+
+    1. Because it is not address specific, we consider it a public event, and send it to everyone who requests it. Basically, we send it encrypted with the first viewing key from any subscription that matches it.
+This has the advantage that the UIs will keep working as expected for existing contracts, but the disadvantage that it might leak unexpected data.   
+    2. We only send it to the `from`. This is more secure, but will impact usability.
+
+My personal preference is for Option 1. Which is in line with the other decisions around not changing the behaviour of public variables, and documenting it for developers.
+
+8. The encrypted event is bubbled out of the enclave on the right subscription channel, and then sent to the wallet extension.
+
+
+### Scenarios
+
+#### 1. Curious user deploys a contract that intends to leak Erc20 balances.
+This is prevented by the protections in place in the ERC20 contract, who will only allow the initiator of a tx to read the balance.
+
+
+#### 2. Curious user listens for all possible events.
+Under Option1. They will get back all events that don't have addresses.
```

Comment:
Maybe make it clear that by you're referring to Option 1 under point 6 above, and not to the first scenario here.